### PR TITLE
Dag model torch

### PR DESCRIPTION
**Pull request by bhuvanmdev**
## Description

This pull request introduces a new model for molecular property prediction, the DAGModel, and includes related changes to support this addition.

### New Model Addition:

* [`deepchem/models/torch_models/dag.py`](diffhunk://#diff-252a858b9c15f5f1642bb005a6f9f41a5563474504af789bf63b0465d3a12fc6R1-R391): Added the `DAGModel` class and its supporting `_DAG` class for molecular property prediction using Directed Acyclic Graphs (DAGs). This includes the implementation of the forward pass, loss functions, and a generator for dataset conversion.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [X] Documentations (modification for documents)

## Checklist

- [X] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [X] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [X] Run `mypy -p deepchem` and check no errors
  - [X] Run `flake8 <modified file> --count` and check no errors
  - [X] Run `python -m doctest <modified file>` and check no errors
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New unit tests pass locally with my changes
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/dag.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+                return [output]
+
+
+class DAGModel(TorchModel):
+    """
+    Directed Acyclic Graph models for molecular property prediction.
+
+    This model is based on the following paper:
+
+    Lusci, Alessandro, Gianluca Pollastri, and Pierre Baldi. "Deep architectures and deep learning in chemoinformatics: the prediction of aqueous solubility for drug-like molecules." Journal of chemical information and modeling 53.7 (2013): 1563-1575.
```
Comment (rbharath) [inline deepchem/models/torch_models/dag.py:171]: This should be in a Reference section

File: deepchem/models/torch_models/tests/test_dag.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+import deepchem as dc
+import tempfile
+from deepchem.trans.transformers import DAGTransformer
+from deepchem.metrics import Metric, roc_auc_score, mean_absolute_error
+from deepchem.molnet import load_bace_classification, load_delaney
+from deepchem.data import NumpyDataset
+import os
+
+
+def get_dataset(mode='classification', featurizer='GraphConv', num_tasks=2):
```
Comment (rbharath) [inline deepchem/models/torch_models/tests/test_dag.py:17]: This is too complicated

Comment (rbharath) [reply deepchem/models/torch_models/tests/test_dag.py:17]: Edit: This is copied directly from TF implementation

File: deepchem/models/torch_models/tests/test_dag.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+                     n_graph_feat=30,
+                     n_outputs=30,
+                     layer_sizes=[100],
+                     layer_sizes_gather=[100],
+                     mode='classification',
+                     n_classes=2,
+                     batch_size=100,
+                     device='cpu')
+
+    def create_weight(shape):
```
Comment (rbharath) [inline deepchem/models/torch_models/tests/test_dag.py:258]: Can you add more comments here explaining?

File: deepchem/models/torch_models/layers.py [hunk 4]
```diff
@@ -6896,7 +6896,7 @@ def __init__(self,
                  activation: str = 'relu',
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:6896]: Let's move these into a small separate PR

**Review by rbharath — COMMENTED** [review]
Not bad, a few small changes needed. One more round should do it

File: deepchem/models/torch_models/tests/test_dag.py (hunk 2 above)
Comment (bhuvanmdev) [reply deepchem/models/torch_models/tests/test_dag.py:17]: Also it DOES NOT use the entire dataset, but rather uses only 20 examples, as mentioned in the `data_points` variable.

**Review by rbharath — COMMENTED** [review]
@JoseAntonioSiguenza Could you do a review on this PR?

**Comment by JoseAntonioSiguenza** [comment]
LGTM

File: deepchem/models/torch_models/dag.py [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    from deepchem.models.torch_models.torch_model import TorchModel
+except (ModuleNotFoundError, ImportError):
+    pass
+
+
+class _DAG(nn.Module):
+    """
+    Directed Acyclic Graph models for molecular property prediction.
+
+    PyTorch implementation of the DAG model described in:
```
Comment (rbharath) [inline deepchem/models/torch_models/dag.py:22]: Please make a References section as described in numpydoc style guide

**Review by rbharath — COMMENTED** [review]
LGTM once CI rerun finishes.
